src/vs/vscode.d.ts

```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
 	}
 
 	/**
+	 * A node provider for the tree explorer contributed by extension.
+	 *
+	 * Providers are registered through (#workspace.registerTreeExplorerNodeProvider) with a
+	 * `providerId` that corresponds to the `treeExplorerNodeProviderId` in the extension's
+	 * `contributes.explorer` section.
+	 *
+	 * The contributed tree explorer will ask the corresponding provider to provide the root
+	 * node and resolve children for each node. In addition, the provider could **optionally**
+	 * provide the following information for each node:
+	 * - label: A human-readable label used for rendering the node.
+	 * - hasChildren: Whether the node has children and is expandable.
+	 * - clickCommand: A command to execute when the node is clicked.
+	 */
+	export interface TreeExplorerNodeProvider<T> {
```

Comment:
@jrieken 

> how does it align with QuickPick

I admit it is very different from QuickPick API. I think it's more aligned to other Providers like `CodeLensProvider`, or `DocumentSymbolProvider`.

> how would a tree'ish quick pick data structure look like

That's what I like about the current design. It doesn't care what your TreeItem data-structure look like. As long as your provider can `getLabel` for your TreeItem, you can work with our API.

> how does it generally fit into the API?

It doesn't resemble any existing API. Previously we did have a version like this:

```ts
export interface TreeExplorerNodeProvider {
    provideRootNode(): Thenable<TreeExplorerNode>;
    resolveChildren(node: TreeExplorerNode): Thenable<TreeExplorerNode[]>;
}

export interface TreeExplorerNode {
    label: string;
    shouldInitiallyExpand: boolean;
    onClickCommand: string;
}
```

but we switched to the `TreeExplorerNode`-less design for reasons @Tyriar mentioned.